File: frontend/src/lib/api/scanHistory.ts
import { apiDelete, apiGet } from './client';
import type { Page } from './chatbot';

export interface ScanHistoryRecord {
  id: string;
  scan_type: 'url_scan' | 'cve_lookup';
  target: string;
  status: 'completed' | 'failed';
  risk_score: number | null;
  severity: 'low' | 'medium' | 'high' | 'critical' | null;
  summary: string;
  details: Record<string, unknown> | null;
  actor: string | null;
  created_at: string;
}

export interface ScanHistoryFilters {
  page?: number;
  pageSize?: number;
  scanType?: 'url_scan' | 'cve_lookup';
  status?: 'completed' | 'failed';
  severity?: 'low' | 'medium' | 'high' | 'critical';
  sort?: 'asc' | 'desc';
}

export function listScanHistory(filters: ScanHistoryFilters = {}): Promise<Page<ScanHistoryRecord>> {
  const params = new URLSearchParams();
  params.set('page', String(filters.page ?? 1));
  params.set('page_size', String(filters.pageSize ?? 20));
  if (filters.scanType) params.set('scan_type', filters.scanType);
  if (filters.status) params.set('status', filters.status);
  if (filters.severity) params.set('severity', filters.severity);
  if (filters.sort) params.set('sort', filters.sort);
  return apiGet<Page<ScanHistoryRecord>>(`/api/tools/scan-history?${params.toString()}`);
}

export function getScanHistoryRecord(id: string): Promise<ScanHistoryRecord> {
  return apiGet<ScanHistoryRecord>(`/api/tools/scan-history/${id}`);
}

export function deleteScanHistoryRecord(id: string): Promise<void> {
  return apiDelete(`/api/tools/scan-history/${id}`);
}

File: backend/database/models/scan_history.py
"""Security scan history.

Only scans that have a meaningful, non-sensitive target are recorded: URL
scans and CVE lookups. **Password checks are never written here** - the
password checker is stateless by design, so there is no row, no target and no
derived fingerprint of the submitted secret anywhere in this table.
"""
import uuid
from typing import Any, Optional

import sqlalchemy as sa
from sqlalchemy.orm import Mapped, mapped_column

from backend.database.base import Base, CreatedAtMixin, JSONVariant, UuidPrimaryKeyMixin, UuidType

SCAN_TYPES = ("url_scan", "cve_lookup")
SCAN_STATUSES = ("completed", "failed")
SEVERITIES = ("low", "medium", "high", "critical")


class SecurityScanRecord(UuidPrimaryKeyMixin, CreatedAtMixin, Base):
    __tablename__ = "security_scan_history"

    scan_type: Mapped[str] = mapped_column(sa.String(32), nullable=False)
    target: Mapped[str] = mapped_column(sa.String(2048), nullable=False)
    status: Mapped[str] = mapped_column(sa.String(16), nullable=False)
    risk_score: Mapped[Optional[int]] = mapped_column(sa.Integer, nullable=True)
    severity: Mapped[Optional[str]] = mapped_column(sa.String(16), nullable=True)
    summary: Mapped[str] = mapped_column(sa.String(500), nullable=False)
    # Structured findings/factors. docs/05_DATA_MODEL.md requires storing the
    # factors behind a risk score, not just the number.
    details: Mapped[Optional[dict[str, Any]]] = mapped_column(JSONVariant, nullable=True)
    actor: Mapped[Optional[str]] = mapped_column(sa.String(128), nullable=True)
    # Ownership: the Supabase Auth user this scan belongs to. Always set from
    # the verified token, never from client input. See conversation.py.
    user_id: Mapped[uuid.UUID] = mapped_column(UuidType, nullable=False)

    __table_args__ = (
        sa.Index("ix_security_scan_history_user_id", "user_id"),
        sa.CheckConstraint(
            "scan_type IN ('url_scan', 'cve_lookup')",
            name="ck_security_scan_history_scan_type",
        ),
        sa.CheckConstraint(
            "status IN ('completed', 'failed')",
            name="ck_security_scan_history_status",
        ),
        sa.CheckConstraint(
            "risk_score IS NULL OR (risk_score >= 0 AND risk_score <= 100)",
            name="ck_security_scan_history_risk_score_range",
        ),
        sa.CheckConstraint(
            "severity IS NULL OR severity IN ('low', 'medium', 'high', 'critical')",
            name="ck_security_scan_history_severity",
        ),
        sa.Index("ix_security_scan_history_created_at", sa.text("created_at DESC")),
        sa.Index("ix_security_scan_history_scan_type", "scan_type"),
        sa.Index("ix_security_scan_history_status", "status"),
    )
